bootstrap/pkg/client/ksapp/ksapp.go

```diff
@@ -0,0 +1,557 @@
+		KsApp: &kstypes.KsApp{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "KsApp",
+				APIVersion: "ksapp.apps.kubeflow.org/v1alpha1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "",
+			},
+			Spec: kstypes.KsAppSpec{
+				Platform:   "none",
+				Version:    "",
+				Repo:       "",
+				Packages:   []string{},
+				Components: []string{},
+				Parameters: map[string][]kstypes.NameValue{},
+			},
+			Status: kstypes.KsAppStatus{
+				Conditions: []kstypes.KsAppCondition{
+					{
+						Type:   "Pending",
+						Status: "Unknown",
+					},
+				},
+			},
+		},
+	}
+	_kfapp.KsApp.Name = options["AppName"].(string)
+	for k, v := range options {
+		x := reflect.ValueOf(_kfapp).Elem().FieldByName(k)
+		x.Set(reflect.ValueOf(v))
+	}
+	if options["KsApp"] == nil {
+		platform := _kfapp.CfgFile.GetString("platform")
+		_kfapp.KsApp.Spec.Platform = platform
+	} else {
+		value := options["KsApp"]
+		ksApp := value.(*kstypes.KsApp)
+		_kfapp.KsApp = ksApp
+	}
+	return _kfapp
+}
+
+func (ksApp *KsApp) writeConfigFile() error {
+	buf, bufErr := yaml.Marshal(ksApp.KsApp)
+	if bufErr != nil {
+		return bufErr
+	}
+	cfgFilePath := filepath.Join(ksApp.AppDir, kftypes.KfConfigFile)
+	cfgFilePathErr := ioutil.WriteFile(cfgFilePath, buf, 0644)
+	if cfgFilePathErr != nil {
+		return cfgFilePathErr
+	}
+	return nil
+}
+
+func (ksApp *KsApp) Apply(resources kftypes.ResourceEnum) error {
+	host, _, err := kftypes.ServerVersion()
+	if err != nil {
+		return fmt.Errorf("couldn't get server version: %v", err)
+	}
+	cli, cliErr := kftypes.GetClientOutOfCluster()
+	if cliErr != nil {
+		return fmt.Errorf("couldn't create client Error: %v", cliErr)
+	}
+	envSetErr := ksApp.envSet(kstypes.KsEnvName, host)
+	if envSetErr != nil {
+		return fmt.Errorf("couldn't create ksonnet env %v Error: %v", kstypes.KsEnvName, envSetErr)
+	}
+	//ks param set application name ${DEPLOYMENT_NAME}
+	name := ksApp.KsApp.Name
+	paramSetErr := ksApp.paramSet("application", "name", name)
+	if paramSetErr != nil {
+		return fmt.Errorf("couldn't set application component's name to %v Error: %v", name, paramSetErr)
+	}
+	namespace := ksApp.KsApp.ObjectMeta.Namespace
+	log.Infof("namespace: %v", namespace)
+	_, nsMissingErr := cli.CoreV1().Namespaces().Get(namespace, metav1.GetOptions{})
+	if nsMissingErr != nil {
+		nsSpec := &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}
+		_, nsErr := cli.CoreV1().Namespaces().Create(nsSpec)
+		if nsErr != nil {
+			return fmt.Errorf("couldn't create namespace %v Error: %v", namespace, nsErr)
+		}
+	}
+	clientConfig, clientConfigErr := kftypes.GetClientConfig()
+	if clientConfigErr != nil {
+		return fmt.Errorf("couldn't load client config Error: %v", clientConfigErr)
+	}
+	applyErr := ksApp.applyComponent([]string{"metacontroller"}, clientConfig)
+	if applyErr != nil {
+		return fmt.Errorf("couldn't create metacontroller component Error: %v", applyErr)
+	}
+	applyErr = ksApp.applyComponent([]string{"application"}, clientConfig)
```

Comment:
We might miss error message from application CRD controller if any components failed to be created.
Still feel more confident to apply each ks component and check its status.